pkg/api/install/install.go

```diff
@@ -19,221 +19,135 @@ limitations under the License.
 package install
 
 import (
-	"fmt"
-
-	"github.com/golang/glog"
-
 	"k8s.io/kubernetes/pkg/api"
-	"k8s.io/kubernetes/pkg/api/meta"
-	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
-	"k8s.io/kubernetes/pkg/apimachinery"
 	"k8s.io/kubernetes/pkg/apimachinery/registered"
 	"k8s.io/kubernetes/pkg/conversion"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util/sets"
-)
-
-const importPrefix = "k8s.io/kubernetes/pkg/api"
 
-var accessor = meta.NewAccessor()
-
-// availableVersions lists all known external versions for this group from most preferred to least preferred
-var availableVersions = []unversioned.GroupVersion{v1.SchemeGroupVersion}
+	// TODO: Remove from here, add to places where this install package is imported.
+	_ "k8s.io/kubernetes/pkg/api/v1/install"
+)
 
 func init() {
-	registered.RegisterVersions(availableVersions)
-	externalVersions := []unversioned.GroupVersion{}
-	for _, v := range availableVersions {
-		if registered.IsAllowedVersion(v) {
-			externalVersions = append(externalVersions, v)
-		}
-	}
-	if len(externalVersions) == 0 {
-		glog.V(4).Infof("No version is registered for group %v", api.GroupName)
-		return
-	}
-
-	if err := registered.EnableVersions(externalVersions...); err != nil {
-		glog.V(4).Infof("%v", err)
-		return
-	}
-	if err := enableVersions(externalVersions); err != nil {
-		glog.V(4).Infof("%v", err)
-		return
-	}
-}
-
-// TODO: enableVersions should be centralized rather than spread in each API
-// group.
-// We can combine registered.RegisterVersions, registered.EnableVersions and
-// registered.RegisterGroup once we have moved enableVersions there.
-func enableVersions(externalVersions []unversioned.GroupVersion) error {
-	addVersionsToScheme(externalVersions...)
-	preferredExternalVersion := externalVersions[0]
-
-	groupMeta := apimachinery.GroupMeta{
-		GroupVersion:  preferredExternalVersion,
-		GroupVersions: externalVersions,
-		RESTMapper:    newRESTMapper(externalVersions),
-		SelfLinker:    runtime.SelfLinker(accessor),
-		InterfacesFor: interfacesFor,
-	}
-
-	if err := registered.RegisterGroup(groupMeta); err != nil {
-		return err
+	if err := registered.AnnounceGroup(&registered.GroupMetaFactory{
```

Comment:
I agree this is net better, I am just constitutionally unable to let
singletons / init side-effects pass without commenting.  We actually have
lots of subtle dependency ordering problems today because import order is
unstable and changes as people refactor code - a list of "Here are all of
the API groups you should use" is infinitely better.

On Wed, May 11, 2016 at 12:39 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/api/install/install.go
> <https://github.com/kubernetes/kubernetes/pull/25434#discussion_r62882318>
> :
>
> > -// registered.RegisterGroup once we have moved enableVersions there.
> > -func enableVersions(externalVersions []unversioned.GroupVersion) error {
> > -	addVersionsToScheme(externalVersions...)
> > -	preferredExternalVersion := externalVersions[0]
> > -
> > -	groupMeta := apimachinery.GroupMeta{
> > -		GroupVersion:  preferredExternalVersion,
> > -		GroupVersions: externalVersions,
> > -		RESTMapper:    newRESTMapper(externalVersions),
> > -		SelfLinker:    runtime.SelfLinker(accessor),
> > -		InterfacesFor: interfacesFor,
> > -	}
> > -
> > -	if err := registered.RegisterGroup(groupMeta); err != nil {
> > -		return err
> > +	if err := registered.AnnounceGroup(&registered.GroupMetaFactory{
>
> Fair enough. One of the problems I'm trying to solve is making the API
> consumable a la carte, and where the parts you don't want don't appear in
> your import tree. I'll see what I can do. There's a ton of low hanging
> fruit here. The immediate inspiration for this change is me being
> embarrassed about the amount of boilerplate people have to copy to make a
> group.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25434/files/0177ec37928214bc7fac744be8935e688181f96a#r62882318>
>